agent/util/archive_tar.go

```diff
@@ -23,117 +23,99 @@ func BuildArchive(ctx context.Context, tarWriter *tar.Writer, rootPath string, i
+				file.Info = symlinkFileInfo
 			}
+		} else {
+			intarball = strings.Replace(file.Path, "\\", "/", -1)
+		}
+		rootPathPrefix := strings.Replace(rootPath, "\\", "/", -1)
+		intarball = strings.Replace(intarball, "\\", "/", -1)
+		intarball = strings.Replace(intarball, rootPathPrefix, "", 1)
+		intarball = filepath.Clean(intarball)
+		intarball = strings.Replace(intarball, "\\", "/", -1)
+
+		//strip any leading slash from the tarball header path
+		intarball = strings.TrimLeft(intarball, "/")
+
+		logger.Infoln("adding to tarball:", intarball)
+		if _, hasKey := processed[intarball]; hasKey {
+			continue
+		} else {
+			processed[intarball] = true
+		}
+		if file.Info.IsDir() {
+			continue
+		}
 
-			_, fileName := filepath.Split(file.Path)
-			for _, ignore := range excludes {
-				if match, _ := filepath.Match(ignore, fileName); match {
-					continue FileChanLoop
-				}
+		_, fileName := filepath.Split(file.Path)
+		for _, ignore := range excludes {
+			if match, _ := filepath.Match(ignore, fileName); match {
+				continue FileLoop
 			}
+		}
 
-			hdr := new(tar.Header)
-			hdr.Name = strings.TrimPrefix(intarball, rootPathPrefix)
-			hdr.Mode = int64(file.Info.Mode())
-			hdr.Size = file.Info.Size()
-			hdr.ModTime = file.Info.ModTime()
+		hdr := new(tar.Header)
+		hdr.Name = strings.TrimPrefix(intarball, rootPathPrefix)
+		hdr.Mode = int64(file.Info.Mode())
+		hdr.Size = file.Info.Size()
+		hdr.ModTime = file.Info.ModTime()
 
-			numFilesArchived++
-			err := tarWriter.WriteHeader(hdr)
-			if err != nil {
-				errChan <- errors.Wrapf(err, "Error writing header for %v", intarball)
-				return
-			}
+		numFilesArchived++
+		err := tarWriter.WriteHeader(hdr)
+		if err != nil {
+			return numFilesArchived, errors.Wrapf(err, "Error writing header for %s", intarball)
+		}
 
-			in, err := os.Open(file.Path)
-			if err != nil {
-				errChan <- errors.Wrapf(err, "Error opening %v", file.Path)
-				return
-			}
-			amountWrote, err := io.Copy(tarWriter, in)
-			if err != nil {
-				logger.Debug(in.Close())
-				errChan <- errors.Wrapf(err, "Error writing into tar for %v", file.Path)
-				return
-			}
+		in, err := os.Open(file.Path)
+		if err != nil {
+			return numFilesArchived, errors.Wrapf(err, "Error opening %s", file.Path)
+		}
+		amountWrote, err := io.Copy(tarWriter, in)
+		if err != nil {
+			logger.Debug(in.Close())
+			return numFilesArchived, errors.Wrapf(err, "Error writing into tar for %s", file.Path)
+		}
 
-			if amountWrote != hdr.Size {
-				logger.Debug(in.Close())
-				errChan <- errors.Errorf(`Error writing to archive for %v:
-					header size %v but wrote %v`,
-					intarball, hdr.Size, amountWrote)
-				return
-			}
+		if amountWrote != hdr.Size {
 			logger.Debug(in.Close())
-			logger.Warning(tarWriter.Flush())
+			return numFilesArchived, errors.Errorf("Error writing to archive for %s: header size %d but wrote %d", intarball, hdr.Size, amountWrote)
 		}
-		done <- true
-	}(pathsToAdd)
-
-	select {
```

Comment:
Is it clear why we were using a channel to begin with? Like, will removing it break anything?

Reply:
These were just so that the 2 goroutines can communicate with each other. Removing the extra goroutine means that these can just be normal variables or removed

Reply:
Ah the question I meant to ask was "why are we using goroutines to begin with" but this still answers it haha